### DISCUSSION TITLE
Add low resolution base64 preview image in image object

### DISCUSSION THREAD

**Discussion by SvenBudak** [directus/directus#19443] [Feature Requests]
### Summary

Hey Team :)

I prefer, in each of my apps, to display a very low-resolution version of the image as a base64 background image on the `<img>` tag while an image is loading. This way, the user can immediately see a rough representation of the image and get an idea of which image is currently being loaded. However, it is necessary for this base64 preview image to be included as a string directly in the data object of the image.

It would also be good if there could be a global setting for this, such as how small the image should be (I usually use 24x24px).

In this video, you can see one of my old apps where I solved this manually. However, it would be nice to have this out of the box in the future.

https://github.com/directus/directus/assets/4930169/c18f5b12-c6ec-4284-aa15-bc57552bc60b

### Basic Example

Lets say we have this image object:
```json
"image": {
    "id": "08677364-62c2-4fe9-8765-498273109c97",
    "preview": "UklGRmwAAABXRUJQVlA4IGAAAAAwBACdASoYABgAP3GuzV+0rSilqAgCkC4JQBmDgsiSpIv92vkkXrAtc0AA/t/8oKpLNnWeC8WeIKaXY//9K5wyClBUuBipO4IHUxFJDApoaH64tJuHJDT40v3ojwAAAAA="
    "storage": "s3",
    "filename_disk": "08677364-62c2-4fe9-8765-498273109c97.jpeg",
    "filename_download": "7e899ca4-7a19-4af9-b3a8-b55d5c63931a.jpeg",
    "title": "7e899ca4 7a19 4af9 B3a8 B55d5c63931a",
    "type": "image/jpeg",
    "folder": null,
    "uploaded_by": "92bdfb3c-2bea-41ce-9cbe-fd002d7d8572",
    "uploaded_on": "2023-08-13T23:28:57.908Z",
    "modified_by": null,
    "modified_on": "2023-08-13T23:28:59.567Z",
    "charset": null,
    "filesize": "77976",
    "width": 700,
    "height": 700,
    "duration": null,
    "embed": null,
    "description": null,
    "location": null,
    "tags": null,
    "metadata": {}
  },
```

then we can use `image.preview` like this in Angular for example:
```
<img [style.background-image]="'url(data:image/png;base64,' + image.preview + ')'" >
```

### Motivation

It is useful for all types of apps/websites. It's simply a part of good UX to show users things like spinners, skeleton loaders, and also preview images.

### Detailed Design

During the upload with Sharp (which you have already integrated), an additional script should simply be created that generates a small (for example 24x24px) base64 string from the uploaded image and saves this into the image object in the database along with the other information.

### Requirements List

Everything is explained already.

### Drawbacks

To allow developers who work with Directus to more easily provide a better UX to the frontend users.

### Alternatives

The alternative would be to realize it again and again with flows. Since a better UX is something that probably every developer wishes for their users, this feature should be implemented by default.

### Adoption Strategy

Existing apps will not be affected, as only a new string is added to the already existing image object.

**Comment by DanielBiegler** [comment]
Edit: 

#19475 mentioned another cool option called Thumbhash which also supports alpha channel and arguably more detail, see here: https://evanw.github.io/thumbhash/

Relevant:

- Thumbhash: https://github.com/evanw/thumbhash
- https://github.com/woltapp/blurhash

**Reply by SvenBudak** [reply]
Thank you! I wasn't familiar with BlurHash; it's exactly what I was asking for here. Thanks for the link!

**Reply by jalada** [reply]
We wrote an extension to add blurhash (similar to Thumbhash) to Directus: https://github.com/pixielabs/directus-extension-blurhash/

**Reply by rijkvanzanten** [reply]
Perfect timing!

**Reply by jamaluddinrumi** [reply]
> We wrote an extension to add blurhash (similar to Thumbhash) to Directus: https://github.com/pixielabs/directus-extension-blurhash/

can confirm it doesn't works with directus@11.1.1

![Screenshot 2024-10-13 at 15 06 01](https://github.com/user-attachments/assets/6e7645c6-1120-4cf3-9a7c-5b5b9960210f)

it even doesn't alter `directus_files` with `blurhash` column specified in migration file (i.e. extension/migrations/202307BLURHASH-add-blurhash.mjs).

**Reply by jalada** [reply]
Not surprised - we made it for Directus 10, well over a year ago!

Open to contributions and suggestions for getting it working with 11.

**Reply by jamaluddinrumi** [reply]
what the only missing here is the migrations folder it should be in the root directory, not in extensions directory.

and then run the database migrate.

voilla it works in directus@11.1.1.

**Reply by jamaluddinrumi** [reply]
![Artboard](https://github.com/user-attachments/assets/7ee15136-2565-4499-9749-95cb7ce6e9b7)

i mean, it's only make the blurhash field is there again.

**Reply by jamaluddinrumi** [reply]
![Screenshot 2024-10-19 at 18 28 52](https://github.com/user-attachments/assets/64d12218-07a7-4f72-9c2d-ef26aef067d1)

yup, it just works! 🥳

**Reply by jamaluddinrumi** [reply]
i mean, it still works.

**Comment by DanielBiegler** [comment]
Technically Directus supports image transforms and presets so you can make small previews already btw.

Using the arguments `?width=100&transforms=[[%22blur%22,%208]]` after your images will create a blurred preview for example. Make it a preset and enjoy. Remember you have the power of [Sharp](https://sharp.pixelplumbing.com/) in the Directus Settings

**Reply by jagu-sayan** [reply]
See https://github.com/transitive-bullshit/lqip-modern

**Reply by rijkvanzanten** [reply]
Sounds like you can achieve exactly what that does as-is! 

> We use sharp to resize input images to a max dimension of 16px and output webp (default) or jpeg images with an encoding quality set to 20. The max dimension is a single, simple variable to tradeoff between encoded image size and visual fidelity.

So set `?width=16&format=webp&quality=20`

**Reply by SvenBudak** [reply]
The key point is that the blurry low-res image must already be contained in the JSON request to ensure an optimal UX experience.

Imagine you are programming an image gallery. You have the following code (in this Angular example):

```html
<ng-container *ngIf="(data$ | async) as data; else loading">
  <ng-container *ngFor="let image of data.gallery; let first = first">
    <img [style.background-image]="'url(data:image/png;base64,' + image.preview + ')'" [src]="image">
  </ng-container>
</ng-container>

<ng-template #loading>
  <div class="w-full h-full flex justify-center items-center">
    <ion-spinner class="w-12 h-12"></ion-spinner>
  </div>
</ng-template>
```

During the initial loading of the view, `(data$ | async) as data; else loading displays loading`. As soon as the `data$` fetch is completed, the UI jumps from loading to `<img [style.background-image]="'url(data:image/png;base64,' + image.preview + ')'" [src]="image">` EXACTLY at the same attosecond, the blurry images should be visible to the user 😅

If the preview image is not delivered with the fetch as `string`, but has to be obtained as an additional extra fetch, there is a risk that the user may briefly see a `blank page` between the `loader` and the `fully loaded images`. In my opinion, this is bad UX. But possibly, I see the whole thing too much through the glasses of a UX/UI designer. Alternatively, I would implement this as a Flow, but I thought others might benefit from it too.

Perhaps this feature could be an activatable feature, which can be set in the project settings along with quality? Since by default you only get image: ID anyway, and you would have to specifically request image.* or image.preview, it would bother the fewest who don't need this feature.

--------

> See https://github.com/transitive-bullshit/lqip-modern

Thanks @jagu-sayan for the link! If I understand it correctly and remember right, Medium (and some other big players) do exactly the same. They either transmit the images in the first data fetch, or even embed the preview image directly as a background image on SSR to the HTML. However, I can't guarantee this statement; I'm not 100% sure if it's true; i just remember that i read years ago alot about this topic. but maybe i remember wrong.

**Reply by rijkvanzanten** [reply]
@SvenBudak One thing to keep in mind though is that base64 text is pretty inefficient (IIRC it's about 33% larger than the binary representation of the same file). For the best UX / performance, I'd argue it might be better to extract the main accent color of the image and return that in the first JSON payload, then use the <img> tag for the tiny size blur placeholder so you can benefit from browser caches, parallelization in the download, and not having to block the main render while waiting for the initial (bigger) JSON payload to download 🤔 (lazy loading)

**Reply by rijkvanzanten** [reply]
Blurhash is an interesting take on that, as they generate a tiny hash for use in the json payload, which then gets expanded to a way bigger data string on the client side in a canvas. The downside of that though is that you're putting that weight on the client side, which might cause some other rendering lag while those hashes are converted to blurry images

**Reply by rijkvanzanten** [reply]
> If the preview image is not delivered with the fetch as string, but has to be obtained as an additional extra fetch, there is a risk that the user may briefly see a blank page between the loader and the fully loaded images

I think that can be solved by properly sizing the image containers, and starting with a background color that contains the predominant color of the image 🙂

**Reply by rijkvanzanten** [reply]
one more thing! (and i'll stop spamming this thread, promise): does medium still do this blurry loading? I just tried inspecting it with the "Slow 3G" network emulation but it just seems to load in the original picture directly instead of displaying a blurry version early 👁️

**Reply by SvenBudak** [reply]
I apologize, I always say base64, but I actually mean that I would leave the exact algorithm to someone who knows about it. I'm not very familiar with it, and I only know base64. Therefore, I use this term very often :)

I would recommend choosing a solution that roughly lets you guess the image. Blurhash is certainly a good approach. It should be considered from the beginning that it might also be sensible to enable the display of large images as a preview. Suppose you have a blog where full-width article teaser images are to be used, 16px image width of course won't be enough to display a 1920x1080p image as a preview.

But even if you only extract the dominant color and send it along, that would be a big step in the right direction :) then I would still have to fall back on my flow solution in the future ;D but here it's not just about my needs :)

**Reply by SvenBudak** [reply]
> one more thing! (and i'll stop spamming this thread, promise): does medium still do this blurry loading? I just tried inspecting it with the "Slow 3G" network emulation but it just seems to load in the original picture directly instead of displaying a blurry version early 👁️

You're right! I assume that they took this feature out when they redesigned Medium (if I remember correctly) a year ago. I remember that (lot time ago) they used this preview blurry image.

**Reply by SvenBudak** [reply]
I once wrote a cloud function in Firebase that made a smaller image from a 500x500 image. The preview string looked like this:
```json
"preview": "UklGRmwAAABXRUJQVlA4IGAAAAAwBACdASoYABgAP3GuzV+0rSilqAgCkC4JQBmDgsiSpIv92vkkXrAtc0AA/t/8oKpLNnWeC8WeIKaXY//9K5wyClBUuBipO4IHUxFJDApoaH64tJuHJDT40v3ojwAAAAA="
```
and this is base64. If there are algorithms that are even more efficient, the string will be significantly smaller. I really don't think that's too much data transfer that gets lost :)

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by NilsBaumgartner1994** [comment]
I suggest writing a hook to create a thumb hash (which is a base64 hash of the image). Thumb hash are a bit cooler: https://evanw.github.io/thumbhash/

**Reply by SvenBudak** [reply]
Daniel has already thrown this into the room: https://github.com/directus/directus/discussions/19443#discussioncomment-6750152

**Comment by antonko** [comment]
This feature is indeed awesome, but there are already a few plugins that handle it really well. For example, here’s my own (apologies for the self-promotion!): https://github.com/antonko/directus-extension-blurhasher

**Comment by NilsBaumgartner1994** [comment]
Nice but you need to restart the server to render the blur hash ? Does not
seem practicable.

Anton Kovalev ***@***.***> schrieb am Mo. 17. Feb. 2025 um
19:51:

**Comment by antonko** [comment]
@NilsBaumgartner1994 

Yes, you do have to restart the server once right after installing or updating the extension. After that, BlurHash is automatically generated for each new upload. Are you suggesting a feature to regenerate all existing BlurHash values without restarting the server?

**Reply by NilsBaumgartner1994** [reply]
No need to recalculate all hashes again.

**Comment by NilsBaumgartner1994** [comment]
Yes, after uploading / creating an image but just for that one image.

Anton Kovalev ***@***.***> schrieb am Mi. 19. Feb. 2025 um
08:23:

**Reply by antonko** [reply]
> Yes, after uploading / creating an image but just for that one image. Anton Kovalev ***@***.***> schrieb am Mi. 19. Feb. 2025 um 08:23:
> […](#)

After installing the extension, you only need to restart Directus us once. From that point on, every newly uploaded image will automatically get its blurhash without any additional restarts required.

**Comment by NilsBaumgartner1994** [comment]
Ah nice, because that Information was missing for me. Nice Job.

Anton Kovalev ***@***.***> schrieb am Mi. 19. Feb. 2025 um
08:41:
